packages/worker/src/connection.ts

```diff
@@ -1,29 +1,208 @@
 import util from 'node:util';
+import { AsyncLocalStorage } from 'node:async_hooks';
+import * as grpc from '@grpc/grpc-js';
+import { getRelativeTimeout } from '@grpc/grpc-js/build/src/deadline';
+import * as proto from 'protobufjs';
+
 import { IllegalStateError } from '@temporalio/common';
-import { Client, Worker, clientUpdateHeaders, TransportError, clientUpdateApiKey } from '@temporalio/core-bridge';
+import {
+  Client,
+  Worker,
+  clientUpdateHeaders,
+  TransportError,
+  clientUpdateApiKey,
+  clientSendRequest,
+} from '@temporalio/core-bridge';
+import { ConnectionLike, Metadata, CallContext, WorkflowService } from '@temporalio/client';
+import { SymbolBasedInstanceOfError } from '@temporalio/common/lib/type-helpers';
 import { NativeConnectionOptions } from './connection-options';
 import { Runtime } from './runtime';
 
 const updateHeaders = util.promisify(clientUpdateHeaders);
 const updateApiKey = util.promisify(clientUpdateApiKey);
 
+@SymbolBasedInstanceOfError('ServiceError')
+export class ServiceError extends Error {
```

Comment:
I don't think we should export this class, nor use the `@SymbolBasedInstanceOfError` decorator on it, as it is likely to be confused with the `ServiceError` class exported by the `client` package (yet, isn't exactly the same thing, as the later corresponds to errors rethrown by the Client, not those coming from the `Connection`).

In the `client` package, we simply use the `isGrpcServiceError()` type guard. So I suggest you keep this class here, just make it private and remove the decorator, and possibly add an `implements GrpcServiceError` if you feel like it.

Reply:
Hmm... I feel like any error that we throw should be exported but I'm not highly opinionated here.
FYI grpc-js `ServiceError` is not a concrete implementation, which is why I created this class.

WDYT?

Reply:
I ended up not exporting this since it's easy to export later while we cannot remove an already exported class.